### PR TITLE
GH-11192 Move non-API OAuth endpoints from api4 to web package

### PR DESCRIPTION
**Pull request by lurcio**
#### Summary
Move non-API OAuth endpoints from api4 to web package

This is pretty much a straight copy of the functions from the `api4` package to `web` so the change is easier to check without altering a load of code in the process.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/11192

### CODE REVIEW

**Comment by lurcio** [comment]
Thanks @reflog, I've updated from master

**Comment by lurcio** [comment]
@jwilander Of course, no problem! Sorry I missed them.

**Comment by jwilander** [comment]
No worries, thanks!

**Comment by lurcio** [comment]
@jwilander To move the `TestAuthorizeOAuthApp`, `TestDeauthorizeOAuthApp`, `TestOAuthAccessToken` and `TestOAuthComplete` tests over to the `web` package I think I need to copy some utility functions from `api4/apitestlib.go`.

Is it best to copy the code, or reference from the `api4` package?

**Comment by lurcio** [comment]
I'm not sure how to proceed here. Taking `TestAuthorizeOAuthApp` as an example, that calls `AdminClient.CreateOAuthApp()` which accesses a route from the `api4` package. As I'm running this test in the `web` package this route isn't available and the test fails.

As `api4` depends on `web` I can't include call out to functions in the `api4` package either.

**Comment by hanzei** [comment]
@jwilander Can you help with some guidance here?

**Comment by jwilander** [comment]
@lurcio what you can do is instead of using the REST API to create the OAuth app, you can create it directly by hitting the app package. For example, replace `AdminClient.CreateOAuthApp()` with `th.App.CreateOAuthApp()` which is in `app/oauth.go`

Does that make sense?

**Comment by lieut-data** [comment]
@lurcio, does @jwilander's comment above help unblock your efforts?

**Comment by lurcio** [comment]
@lieut-data yes, I think it does. I'll have a look at it tonight or tomorrow. Thank you @jwilander!

**Comment by jasonblais** [comment]
@lurcio Thanks for your contribution! Friendly reminder about the previous feedback, let us know if we can help with any questions :)

**Comment by lurcio** [comment]
@jasonblais apologies, I made the mistake of trying to add all tests at once and ended up with a whole bunch of problems. I'll try adding one at a time, fix it totally and move on to the next.

I have just moved `TestAuthorizeOAuthApp` over, and am running into an error when I run it:

`.../go/src/github.com/mattermost/mattermost-server/web/oauth_test.go:72: Expected no error, got ": Invalid or expired session, please login again., "`

I think the test requires a user to be logged in for it to run. However, when I try to do this I think that the `/users/login` route isn't available.

**Comment by jasonblais** [comment]
Thanks @lurcio!

@jwilander pinging you to help with the above question, thanks!

**Comment by jwilander** [comment]
@lurcio you can use the app layer to create a session from which you can use the token to authenticate

You can use this method https://github.com/mattermost/mattermost-server/blob/master/app/session.go#L14

Passing it a session struct, created similar to this https://github.com/mattermost/mattermost-server/blob/master/app/login.go#L127 (leaving out DeviceId)

Let me know if that doesn't work and I can help out some more

**Comment by lurcio** [comment]
Thank you @jwilander! I can use `CreateSession` as you suggested and get a session with a token set. I'm afraid I'm drawing a blank on how to authenticate requests using that. May I have another nudge in the right direction please?

**Comment by jwilander** [comment]
You should be able to set `AuthToken` on the client to the session token and `AuthType` to model.HEADER_BEARER

https://github.com/mattermost/mattermost-server/blob/master/model/client4.go#L61

**Comment by lurcio** [comment]
Thank you for your help @jwilander! The tests mostly pass now, however I have commented out a few sections (marked with `TODO: This fails...`). These are parts that pull information from the API and I have yet to find a way to work around this.

**Comment by hanzei** [comment]
@lurcio Can we help with anything to get the last tests running?

**Comment by jwilander** [comment]
Sorry for the late response again. Those are pretty critical parts of the tests because they are testing to confirm if the tokens work.

Hmmm, we can't init the API to test these because the `api4` package depends on the `web` package. I see that we could do one of two things:

1. Use one of the endpoints in the `web` package that requires authentication to test the tokens (might be hard because none of them are simple GETs), or
2. Create a test endpoint that only gets initialized during the tests (maybe in the setup of web_test.go) that we use to test the tokens

What do you think @lurcio ?

**Comment by lurcio** [comment]
@jwilander sorry, I have just returned back from vacation. Option 2 sounds best, but I think I'm a bit out of my depth. I'm not sure what the new endpoint should do - some calls to `GetUsers` work, but others fail. Should I move all calls like that to a new endpoint? How are tokens tested/validated?

**Comment by jwilander** [comment]
I think the new endpoint doesn't need to do anything other than return 200 if it's successful. If you hook it up like the other endpoints using `ApiSessionRequired` handler, the token handling etc. that we want to test will be auto set up for the endpoint. Then in the tests you make requests against it, making sure you get a 200 OK with a valid OAuth token or a 401 Unauthorized when you don't

**Comment by lurcio** [comment]
@jwilander OK, I think I have run into a problem with option 2. There is an `/{anything:.*}` route defined in `static.go` which takes precedence over any new routes I define in `web_test.go` and swallows up any API requests. 😞 

Adding it somewhere else like `oauth.go` should work, but then it exists outside of tests.

**Comment by lurcio** [comment]
I have added the route to `oauth.go` and it seems to work there, but it's available outside of tests.

**Comment by jwilander** [comment]
@lurcio I think that's totally fine 👍  We have precedent for testing endpoints done like this in the api4 package.

**Comment by jwilander** [comment]
I'll review shortly

File: web/oauth.go [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+	w.MainRouter.Handle("/oauth/{service:[A-Za-z0-9]+}/complete", w.ApiHandler(completeOAuth)).Methods("GET")
+	w.MainRouter.Handle("/oauth/{service:[A-Za-z0-9]+}/login", w.ApiHandler(loginWithOAuth)).Methods("GET")
+	w.MainRouter.Handle("/oauth/{service:[A-Za-z0-9]+}/mobile_login", w.ApiHandler(mobileLoginWithOAuth)).Methods("GET")
+	w.MainRouter.Handle("/oauth/{service:[A-Za-z0-9]+}/signup", w.ApiHandler(signupWithOAuth)).Methods("GET")
+
+	// Old endpoints for backwards compatibility, needed to not break SSO for any old setups
+	w.MainRouter.Handle("/api/v3/oauth/{service:[A-Za-z0-9]+}/complete", w.ApiHandler(completeOAuth)).Methods("GET")
+	w.MainRouter.Handle("/signup/{service:[A-Za-z0-9]+}/complete", w.ApiHandler(completeOAuth)).Methods("GET")
+	w.MainRouter.Handle("/login/{service:[A-Za-z0-9]+}/complete", w.ApiHandler(completeOAuth)).Methods("GET")
+	w.MainRouter.Handle("/api/v4/test", w.ApiSessionRequired(testHandler)).Methods("GET")
```
Comment (jwilander) [inline web/oauth.go]: Could we actually change this to `/api/v4/oauth_test` so it's a bit clearer what it's being used for? `/api/v4/test` is a bit too generic IMO

Comment (lurcio) [reply web/oauth.go]: Of course! Done.

**Comment by lurcio** [comment]
Wow, thanks for all of your help @jwilander!

**Pull request merged**